src/typecheck.rs

```diff
@@ -1,42 +1,215 @@
+                facts.push(if head == &"value-eq".into() {
+                    assert!(args.len() == 2);
+                    NormFact::ConstrainEq(args[0], args[1])
+                } else {
+                    let (out, args) = args.split_last().unwrap();
+                    if egraph.functions.contains_key(head) {
+                        NormFact::Compute(*out, NormExpr::Call(*head, args.to_vec()))
+                    } else {
+                        NormFact::Assign(*out, NormExpr::Call(*head, args.to_vec()))
+                    }
+                });
+            }
+            facts
+        }
+
+        let UnresolvedCoreRule { head, body } = self;
+        let Actions(head) = head.clone();
+        NormRule {
+            head,
+            body: to_fact(body, egraph),
+        }
+    }
+}
+
+pub(crate) fn facts_to_query(body: &Vec<NormFact>, typeinfo: &TypeInfo) -> Query<Symbol> {
+    fn to_atom_term(s: Symbol, typeinfo: &TypeInfo) -> AtomTerm {
+        if typeinfo.global_types.contains_key(&s) {
+            AtomTerm::Global(s)
+        } else {
+            AtomTerm::Var(s)
+        }
+    }
+    let mut atoms = vec![];
+    for fact in body {
+        match fact {
+            NormFact::Assign(symbol, NormExpr::Call(head, args))
+            | NormFact::Compute(symbol, NormExpr::Call(head, args)) => {
+                let args = args
+                    .iter()
+                    .chain(once(symbol))
+                    .cloned()
+                    .map(|s| to_atom_term(s, typeinfo))
+                    .collect();
+                let head = *head;
+                atoms.push(Atom { head, args });
+            }
+            NormFact::AssignVar(lhs, rhs) => atoms.push(Atom {
+                head: "value-eq".into(),
+                args: vec![
+                    to_atom_term(*lhs, typeinfo),
+                    to_atom_term(*rhs, typeinfo),
+                    AtomTerm::Literal(Literal::Unit),
+                ],
+            }),
+            NormFact::ConstrainEq(lhs, rhs) => atoms.push(Atom {
+                head: "value-eq".into(),
+                args: vec![
+                    to_atom_term(*lhs, typeinfo),
+                    to_atom_term(*rhs, typeinfo),
+                    AtomTerm::Literal(Literal::Unit),
+                ],
+            }),
+            NormFact::AssignLit(symbol, lit) => atoms.push(Atom {
+                head: "value-eq".into(),
+                args: vec![
+                    to_atom_term(*symbol, typeinfo),
+                    AtomTerm::Literal(lit.clone()),
+                    AtomTerm::Literal(Literal::Unit),
+                ],
+            }),
+        }
+    }
+    Query { atoms }
+}
+
+impl NormRule {
+    // keep it here while it's untested
+    #![allow(unused)]
+    pub(crate) fn to_core_rule(&self, typeinfo: &TypeInfo) -> UnresolvedCoreRule {
```

Comment:
Why is it unused? I thought the current pipeline required it

Reply:
Good catch. It is indeed used in our pipeline. I removed the `allow(unused)` annotation.